src/dinit-iostream.cc

```diff
@@ -0,0 +1,740 @@
+            iovs_to_write = 2;
+            bufv[1].iov_base = buf->get_buf_base();
+            bufv[1].iov_len = buf->get_length() - len;
+        }
+
+        ssize_t r = bp_sys::writev(fd, bufv, iovs_to_write);
+        if (r < 0) {
+            io_error = errno;
+            return false;
+        }
+        buf->consume(r);
+    }
+
+    return true;
+}
+
+bool ostream::flush()
+{
+    bool r = flush_nx();
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const char *msg) noexcept
+{
+    size_t len = strlen(msg);
+    return (put(msg, len) == static_cast<ssize_t>(len));
+}
+
+bool ostream::write(const char *msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(char msg) noexcept
+{
+    return (put(&msg, 1) == 1);
+}
+
+bool ostream::write(char msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const std::string &msg) noexcept
+{
+    return (put(msg.c_str(), msg.size()) == static_cast<ssize_t>(msg.size()));
+}
+
+bool ostream::write(const std::string &msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const int num) noexcept
+{
+    char tmp[INT_MAX_CHARS + 1];
+    int len = snprintf(tmp, INT_MAX_CHARS + 1, "%d", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const int num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const unsigned num) noexcept
+{
+    char tmp[UINT_MAX_CHARS + 1];
+    int len = snprintf(tmp, UINT_MAX_CHARS + 1, "%d", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const unsigned num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const long num) noexcept
+{
+    char tmp[LONG_MAX_CHARS + 1];
+    int len = snprintf(tmp, LONG_MAX_CHARS + 1, "%ld", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const long num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
```

Comment:
~This will definitely throw an exception~
failed write leaves either buffer_failbit or io_failbit set